File: python/00004_median_of_two_sorted_arays.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # Ensure nums1 is the smaller array
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        m, n = len(nums1), len(nums2)
        total_left = (m + n + 1) // 2

        left, right = 0, m

        while left <= right:
            partitionA = (left + right) // 2
            partitionB = total_left - partitionA

            maxLeftA = float('-inf') if partitionA == 0 else nums1[partitionA - 1]
            minRightA = float('inf') if partitionA == m else nums1[partitionA]

            maxLeftB = float('-inf') if partitionB == 0 else nums2[partitionB - 1]
            minRightB = float('inf') if partitionB == n else nums2[partitionB]

            # Check if we found the correct partition
            if maxLeftA <= minRightB and maxLeftB <= minRightA:
                if (m + n) % 2 == 1:
                    return max(maxLeftA, maxLeftB)
                else:
                    return (max(maxLeftA, maxLeftB) + min(minRightA, minRightB)) / 2.0
            elif maxLeftA > minRightB:
                right = partitionA - 1
            else:
                left = partitionA + 1

        raise ValueError("Input arrays are not sorted.")

# Time complexity - O(log(min(m, n)))
# Space complexity - O(1)

File: java/00004_median_of_two_sorted_arays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // Make sure nums1 is the smaller array to reduce binary search range
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length;
        int n = nums2.length;
        int totalLeft = (m + n + 1) / 2; // Number of elements in the left partition

        int left = 0;
        int right = m; // Search space in nums1

        while (left <= right) {
            int partitionA = (left + right) / 2;
            int partitionB = totalLeft - partitionA; // Remaining elements go to nums2

            // If partition is at index 0, maxLeft is -∞ (no elements on left)
            int maxLeftA = (partitionA == 0) ? Integer.MIN_VALUE : nums1[partitionA - 1];
            int minRightA = (partitionA == m) ? Integer.MAX_VALUE : nums1[partitionA];

            int maxLeftB = (partitionB == 0) ? Integer.MIN_VALUE : nums2[partitionB - 1];
            int minRightB = (partitionB == n) ? Integer.MAX_VALUE : nums2[partitionB];

            // Check if we have the perfect partition
            if (maxLeftA <= minRightB && maxLeftB <= minRightA) {
                // If total length is odd, median is max of left side
                if ((m + n) % 2 == 1) {
                    return Math.max(maxLeftA, maxLeftB);
                }
                // If even, median is average of maxLeft and minRight
                return (Math.max(maxLeftA, maxLeftB) + Math.min(minRightA, minRightB)) / 2.0;
            }
            // Move search space
            else if (maxLeftA > minRightB) {
                right = partitionA - 1; // Move left
            } else {
                left = partitionA + 1; // Move right
            }
        }

        throw new IllegalArgumentException("Input arrays are not sorted.");
    }
}

// Time complexity - O(log(min(m, n)))
// Space complexity - O(1)

